build/gengo.sh

```diff
@@ -0,0 +1,150 @@
+  exit 1
+fi
+
+# Main logic which runs the gengo tool on packages requesting generation.
+
+# Set DEBUG=1 for extra info and checks
+DEBUG=
+# location of the prebuilt generator
+dcg="$PWD/$tool"
+
+# original pwd
+O=$PWD
+
+# split each GOPATH entry by :
+gopaths=($(IFS=: ; echo $GOPATH))
+# The first GOPATH entry is where go_genrule puts input source files.
+# So use this one
+srcpath="${gopaths[0]}"
+
+# Use the go version bazel wants us to use, not system path
+GO="$GOROOT/bin/go"
+
+# Find all packages that need generation
+files=()
+#   first, all packages with the tag, except the tool package itself
+white=($(find . -name *.go | \
+    grep -v "${tool##*/}" | \
+    (xargs grep -l "$match" || true)))
+#   now, munge these names a bit:
+#     rename /staging/src to /vendor
+#     change ./ to k8s.io/kubernetes
+#     ignore packages that only have +k8s:foo-gen=false lines
+dirs=()
+for w in "${white[@]}"; do
+  if grep "$match" "$w" | grep -q -v "${match}false"; then
+    dirs+=("$(dirname "$w" | sed -e 's|./staging/src/|./vendor/|;s|^./|k8s.io/kubernetes/|')")
+  elif [[ -n "${DEBUG:-}" ]]; then
+    echo SKIP: $w, has only ${match}false tags
+  fi
+done
+
+if [[ -n "${DEBUG:-}" ]]; then
+  echo "Generating: $(IFS=$'\n ' ; echo "${dirs[*]}" | sort -u)"
+else
+  echo "Generating $out for ${#dirs} packages..."
+fi
+# Send the tool the comma-separate list of packages to generate
+packages="$(IFS="," ; echo "${dirs[*]}")"
+# Run the tool
+$dcg \
+  -v 1 \
+  -i "$packages" \
+  -O "${out%%???}" \
+  -h "$head" \
+  $tool_flags
+
+# DEBUG: Ensure we generated each file
+# TODO(fejta): consider deleting this
+if [[ -n "${DEBUG:-}" ]]; then
+  for p in "${dirs[@]}"; do
+    found=false
+    for s in "" "k8s.io/kubernetes/vendor/" "staging/src/"; do
+      if [[ -f "$srcpath/src/$s$p/$out" ]]; then
+        found=true
+        if [[ $s == "k8s.io/kubernetes/vendor/" ]]; then
+          grep -A 1 import $srcpath/src/$s$p/$out
+        fi
+        break
+      fi
+    done
+    if [[ $found == false ]]; then
+      echo FAILED: $p
+      failed=1
+    else
+      echo FOUND: $p
+    fi
+  done
+  if [[ -n "${failed:-}" ]]; then
+    exit 1
+  fi
+fi
+
+oifs="$IFS"
+IFS=$'\n'
+# use go list to create lines of pkg import import ...
+for line in $($GO list -f '{{.ImportPath}}{{range .Imports}} {{.}}{{end}}' "${dirs[@]}"); do
+  pkg="${line%% *}" # first word
+  imports="${line#* }" # everything after the first word
+  IFS=' '
+  deps="$srcpath/src/$pkg/$out.deps"
+  echo > "$deps"
+  for dep in $imports; do
+    if [[ ! "$dep" == k8s.io/kubernetes/* ]]; then
+      continue
+    fi
+    echo //${dep#k8s.io/kubernetes/}:go_default_library >> $deps
+  done
+  IFS=$'\n'
+done
+IFS="$oifs"
+
+# copy $1 to where bazel expects to find it
+move-out() {
+  D="$(OUTS)"
+  dst="$O/${D%%/genfiles/*}/genfiles/build/$1/$out"
+  options=(
+    "k8s.io/kubernetes/$1"
+    "${1/vendor\//}"
+    "k8s.io/kubernetes/staging/src/${1/vendor\//}"
+  )
+  found=false
```

Comment:
maybe steal the [method from k/k `hack/lib/util.sh`](https://github.com/kubernetes/kubernetes/blob/a743392937c022a87831227ef152471359fd8d9e/hack/lib/util.sh#L142-L161), which is only one line?
```
local -r src=$( (ls -t "${options[@]}" 2>/dev/null || true) | head -1 )
```